hazelcast/src/test/java/com/hazelcast/client/heartbeat/ClientHeartbeatTest.java

```diff
@@ -324,4 +328,63 @@ public boolean decodeRemoveResponse(ClientMessage clientMessage) {
             }
         };
     }
+
+    @Test
+    public void testClientMembershipEvents_onSplitBrain() {
+        Config config = new Config();
+        HazelcastInstance instanceA = hazelcastFactory.newHazelcastInstance(config);
+
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig()
+                .setClusterConnectTimeoutMillis(Integer.MAX_VALUE);
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+
+        HazelcastInstance instanceB = hazelcastFactory.newHazelcastInstance(config);
+        HazelcastInstance instanceC = hazelcastFactory.newHazelcastInstance(config);
+
+        Member memberA = instanceA.getCluster().getLocalMember();
+        Member memberB = instanceB.getCluster().getLocalMember();
+        Member memberC = instanceC.getCluster().getLocalMember();
+
+        CountDownLatch splitLatch = new CountDownLatch(1);
+        CountDownLatch bRemovedLatch = new CountDownLatch(1);
+        CountDownLatch aRemovedLatch = new CountDownLatch(1);
+
+        CountDownLatch switchedToCLatch = new CountDownLatch(1);
+        client.getCluster().addMembershipListener(new MembershipListener() {
+            @Override
+            public void memberAdded(MembershipEvent membershipEvent) {
+                if (memberC.equals(membershipEvent.getMember())) {
+                    switchedToCLatch.countDown();
+                }
+            }
+
+            @Override
+            public void memberRemoved(MembershipEvent membershipEvent) {
+                if (memberC.equals(membershipEvent.getMember())) {
+                    splitLatch.countDown();
+                }
+
+                if (memberB.equals(membershipEvent.getMember())) {
+                    bRemovedLatch.countDown();
+                }
+
+                if (memberA.equals(membershipEvent.getMember())) {
+                    aRemovedLatch.countDown();
+                }
+            }
+        });
+
+        blockCommunicationBetween(instanceA, instanceC);
+        closeConnectionBetween(instanceA, instanceC);
+        blockCommunicationBetween(instanceB, instanceC);
+        closeConnectionBetween(instanceB, instanceC);
+
+        assertOpenEventually(" A B | C", splitLatch);
+        instanceB.shutdown();
+        assertOpenEventually(" A | C", bRemovedLatch);
+        instanceA.shutdown();
+
```

Comment:
Minor: we don't check whether or not `aRemovedLatch` is open

Reply:
Good catch 👍